multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/util/ProcessTimeCalculator.java

```diff
@@ -64,24 +63,46 @@ private long calculateFilteredProcessActivitiesTime(List<HistoricActivityInstanc
                                                         Predicate<HistoricActivityInstance> filter) {
         return processActivities.stream()
                                 .filter(filter)
-                                .mapToLong(HistoricActivityInstance::getDurationInMillis)
+                                .mapToLong(this::calculateActivityDuration)
                                 .sum();
     }
 
-    private boolean isCallActivity(HistoricActivityInstance historicActivityInstance) {
-        return CALL_ACTIVITY_TYPE.equals(historicActivityInstance.getActivityType());
+    // unfortunately, HistoricActivityInstance and HistoricProcessInstance have no common ancestor
+    // even though both have getStart/EndTime methods
+    // that's why there is duplication in the methods
+    private long calculateActivityDuration(HistoricActivityInstance activityInstance) {
+        Date startTime = activityInstance.getStartTime();
+        Date endTime = determineProcessActivityEndTime(activityInstance);
+        return endTime.getTime() - startTime.getTime();
+    }
+
+    private Date determineProcessActivityEndTime(HistoricActivityInstance activityInstance) {
+        return activityInstance.getEndTime() == null ? new Date(currentTimeSupplier.getAsLong()) : activityInstance.getEndTime();
     }
 
     private boolean isTimerEvent(HistoricActivityInstance historicActivityInstance) {
         return TIMER_EVENT_TYPE.equals(historicActivityInstance.getActivityType());
     }
 
-    @Immutable
-    @JsonSerialize(as = ImmutableProcessTime.class)
-    @JsonDeserialize(as = ImmutableProcessTime.class)
-    public interface ProcessTime {
-        long getProcessDuration();
+    private boolean isSequenceFlow(HistoricActivityInstance historicActivityInstance) {
+        return SEQUENCE_FLOW_TYPE.equals(historicActivityInstance.getActivityType());
+    }
+
+    private boolean isGateway(HistoricActivityInstance historicActivityInstance) {
+        String activityType = historicActivityInstance.getActivityType();
+        return activityType != null && activityType.endsWith(GATEWAY_TYPE);
+    }
 
-        long getDelayBetweenSteps();
+    private long getDelayBetweenActivities(List<HistoricActivityInstance> activities) {
+        HistoricActivityInstance[] processActivities = activities.toArray(new HistoricActivityInstance[0]);
+        Arrays.sort(processActivities, Comparator.comparing(HistoricActivityInstance::getStartTime));
+        long result = 0;
+
+        for (int i = 0; i < processActivities.length - 1; i++) {
+            long delay = processActivities[i + 1].getStartTime()
+                                                 .getTime() - determineProcessActivityEndTime(processActivities[i]).getTime();
+            result += delay < 0 ? 0 : delay;
```

Comment:
Add a comment here saying that this is done, because of the possibility of parallel call activities.

Reply:
Done